src/Plutarch/Test/QuickCheck/Instances.hs

```diff
@@ -0,0 +1,974 @@
+    CoArbitrary (coarbitrary),
+    Function (function),
+    Gen,
+    Testable (property),
+    chooseInt,
+    elements,
+    frequency,
+    functionMap,
+    sized,
+    variant,
+    vectorOf,
+ )
+
+-- | @since 2.0.0
+instance Testable (TestableTerm PBool) where
+    property (TestableTerm t) = property (plift t)
+
+{- | TestableTerm is a wrapper for closed Plutarch term. This
+     abstraction allows Plutarch values can be generated via QuickCheck
+     Generator.
+
+     Hint: Typechecker is picky about how TestableTerm is constructed.
+     Meaning, TestableTerm throw error when it's composed.
+
+ @since 2.0.0
+-}
+data TestableTerm a = TestableTerm {unTestableTerm :: forall s. Term s a}
+
+-- | @since 2.0.0
+liftTestable ::
+    forall (a :: S -> Type) (b :: S -> Type).
+    (forall (s :: S). Term s a -> Term s b) ->
+    TestableTerm a ->
+    TestableTerm b
+liftTestable f (TestableTerm x) = TestableTerm $ f x
+
+-- | @since 2.0.0
+lift2Testable ::
+    forall (a :: S -> Type) (b :: S -> Type) (c :: S -> Type).
+    (forall (s :: S). Term s a -> Term s b -> Term s c) ->
+    TestableTerm a ->
+    TestableTerm b ->
+    TestableTerm c
+lift2Testable f (TestableTerm x) (TestableTerm y) = TestableTerm $ f x y
+
+-- | @since 2.0.0
+instance (forall (s :: S). Num (Term s a)) => Num (TestableTerm a) where
+    (+) = lift2Testable (+)
+    (*) = lift2Testable (*)
+    abs = liftTestable abs
+    negate = liftTestable negate
+    signum = liftTestable signum
+    fromInteger i = TestableTerm (fromInteger i :: Term s a)
+
+-- | @since 2.0.0
+instance (forall (s :: S). Eq (Term s a)) => Eq (TestableTerm a) where
+    (TestableTerm x) == (TestableTerm y) = x == y
+
+{- | For any Plutarch Type that have `PShow` instance, `Show` is
+     available as well. Unfortunately, for those that doesn't have
+     @PShow@, `forAllShow` with custom show function is required to
+     execute property check.
+
+     This only works with those that have PShow instance; it may cause
+     problems when dealing with Plutarch.Api datatypes which does not
+     have PShow instances.
+
+ @since 2.0.0
+-}
+instance PShow a => Show (TestableTerm a) where
+    show (TestableTerm term) =
+        case compile (Config{tracingMode = DoTracing}) $ ptraceError (pshow term) of
+            Left err -> show err
+            Right (evalScript -> (_, _, trace)) ->
+                T.unpack . T.intercalate " " $ trace
+
+{- | PArbitrary is Plutarch equivalent of `Arbitrary` typeclass from
+     QuickCheck. It generates randomized closed term, which can be used
+     to test property over Plutarch code without compiling and evaluating.
+
+     Default implmentation is given for any Plutarch types that
+     implments @PLift a@ and @Arbitrary (PLifted a)@. This will Generate
+     a haskell value and convert that into Plutarch term using `pconstant`.
+
+     Other more complex Plutarch types, like `PMaybe`, requires mannual
+     implmentation.
+
+ @since 2.0.0
+-}
```

Comment:
"is the Plutarch equivalent of the 'Arbitrary' type class from QuickCheck."
'It generates pseudo-random closed terms, which can be used to test properties over Plutarch code without having to compile and evaluate.'
"Default implementations are given for any Plutarch type that implements 'PLift' and 'Arbitrary' for its 'PLifted' equivalent. This generates a Haskell value, then converts it into a Plutarch term using 'pconstant'."

The last paragraph can go.